src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -344,6 +362,89 @@ void AppendRestoreCode(MachineCode& trampoline) {
   return outcome::success();
 }
 
+// First backup all the (potential) return values - compare section "3.2.3 Parameter Passing" in
+// "System V Application Binary Interface"
+// https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.99.pdf. Then call the exit payload, restore
+// the return values and finally jump the actual return address.
+void AppendCallToExitPayloadAndJumpToReturnAddress(uint64_t exit_payload_function_address,
+                                                   MachineCode& return_trampoline) {
+  // Backup rax, rdx, st(0), st(1).
+  // push rax                                        50
+  // push rdx                                        52
+  // sub $0x0a, rsp                                  48 83 ec 0a
+  // fstpt (rsp)                                     db 3c 24
+  // sub $0x0a, rsp                                  48 83 ec 0a
+  // fstpt (rsp)                                     db 3c 24
+  return_trampoline.AppendBytes({0x50})
+      .AppendBytes({0x52})
+      .AppendBytes({0x48, 0x83, 0xec, 0x0a})
+      .AppendBytes({0xdb, 0x3c, 0x24})
+      .AppendBytes({0x48, 0x83, 0xec, 0x0a})
+      .AppendBytes({0xdb, 0x3c, 0x24});
+
+  // We align the stack to 32 bytes first: round down to a multiple of 32, subtract another 24 and
+  // then push 8 byte original rsp. So we are 32 byte aligned after these commands and we can 'pop
+  // rsp' later to undo this.
+  // mov rax, rsp                                    48 89 e0
+  // and rsp, $0xffffffffffffffe0                    48 83 e4 e0
+  // sub rsp, 0x18                                   48 83 ec 18
+  // push rax                                        50
+  return_trampoline.AppendBytes({0x48, 0x89, 0xe0})
+      .AppendBytes({0x48, 0x83, 0xe4, 0xe0})
+      .AppendBytes({0x48, 0x83, 0xec, 0x18})
+      .AppendBytes({0x50});
+
+  // Store xmm0 and xmm1 on the stack.
+  // sub     rsp, 16                                 48 83 ec 10
```

Comment:
nit. Should we `sub rsp, 32` once instead of doing `sub rsp, 16` twice? Maybe there is no real benefit, not sure.

Reply:
Then the first movdqa would need to get an offset:
```
sub rsp, 32
movdqa 16+(rsp), xmm0
movdqa (rsp), xmm1
```
Sure, this is equivalent, but I like this better since it looks more like 
```
push xmm0
push xmm1
```
(but there is no push for xmm* ...)

Reply:
yep that's right we'd need the offset, not sure which is faster but I think it's fine as is. Thanks for the additional info.